Implement song queue on frontend

parthivshaji opened the pull request:
# 🚀 Pull Request

## 📋 Summary
<!-- Provide a brief explanation of the changes. What problem does this PR fix? -->
Adds queue functionality to song player.
- Implemented drag/drop functionality
- Allows user to delete queue items
- User can add new songs in songplayerpage using bottom input box
- Playlist component is axed, loading playlist just adds all songs to queue
- Preloads next two songs in queue on halfway point of current song
- Songs that are loaded have genius image thumbnail in queue

## 🔍 Related Issues
<!-- Link to related issues, e.g., "Fixes #123" or "Closes #456" -->
Fixes #142, 
Fixes #156
Fixes #31

## 📷 Screenshots (if applicable)
<!-- Add screenshots or GIFs if visual changes were made. -->
![image](https://github.com/user-attachments/assets/c87414fa-6a82-406f-a722-1ea3c2040cfe)
![image](https://github.com/user-attachments/assets/97a7b0fc-622f-4a1b-aeab-f485163f72fa)

## 💬 Additional Comments
<!-- Any additional context or thoughts? -->

darint916 commented on frontend/lyric-sync-app/src/components/DraggableQueueItem/DraggableQueueItem.tsx:
```diff
@@ -0,0 +1,167 @@
+      <ListItemButton 
+        onClick={() => !hasError && onPlay(item)}
+        className={`queue-item ${isCurrentSong ? 'active-song' : ''} ${hasError ? 'error-item' : ''}`}
+        sx={{
```
make sure the rest of the styling in the doc is also consistent

darint916 commented on frontend/lyric-sync-app/src/components/DraggableQueueItem/DraggableQueueItem.tsx:
```diff
@@ -0,0 +1,167 @@
+            src={songImageUrl} 
+            variant="square"
+            sx={{ 
+              width: 40, 
```
more in line styling... also neaten up the format pls

darint916 commented on frontend/lyric-sync-app/src/contexts/SongContext.ts:
```diff
@@ -1,13 +1,18 @@
   setSong: (_song: FullSongResponseDto) => void;
   setQuizQuestions: (_quiz: QuizQuestionDto[]) => void;
+  queue: QueueItem[];
+  setQueue: (_queue: QueueItem[] | ((_prevQueue: QueueItem[]) => QueueItem[])) => void;
```
why are we denoting with underscores?

parthivshaji replied:
Otherwise it gets flagged as unused var by linter, and we need this placeholder variable for typescript syntax

darint916 commented on frontend/lyric-sync-app/src/pages/HomePage/HomePage.tsx:
```diff
@@ -14,36 +14,112 @@ import { useNavigate } from 'react-router-dom';
+
+  const handleLoadPlaylist = async () => {
+    if (!playlistUrl.trim()) {
+      setError('Please enter a valid YouTube playlist URL.');
```
Should say Invalid Youtube Playlist URL Entered

darint916 commented on frontend/lyric-sync-app/src/pages/HomePage/HomePage.tsx:
```diff
@@ -14,36 +14,112 @@ import { useNavigate } from 'react-router-dom';
+
+    try {
+      const playlistId = extractPlaylistId(playlistUrl);
+      if (!playlistId) {throw new Error('Invalid YouTube playlist URL');}
```
if Invalid, should go next in the playlist automatically?

darint916 replied:
Maybe have a quick popup in the q that should show like it errored out, maybe set that q background for the vid a diff color

parthivshaji replied:
Leaving as is, similar to YouTube queue

darint916 commented on frontend/lyric-sync-app/src/pages/HomePage/HomePage.tsx:
```diff
@@ -14,36 +14,112 @@ import { useNavigate } from 'react-router-dom';
+        title: video.title || 'Unknown Track',
+        artist: video.artist || 'Unknown Artist',
+        youTubeId: extractYouTubeVideoId(video.url) || '',
+        lyrics: '',
```
dont we have multiple sets of lyrics? (translated, romanized. etc

parthivshaji replied:
This is for the lyrics parameter in FullSongRequestDTO

darint916 commented on frontend/lyric-sync-app/src/pages/HomePage/HomePage.tsx:
do we really want playlist viewer in the homepage file, shouldn't this be abstracted into another page, not sure? but this feels like it doesn't fit in with home page in terms of feature separation. Maybe have a queue component tsx

parthivshaji replied:
I agree with queue component tsx, but playlist datagrid was axed, and the remaining code was just one handleLoadPlaylist function so thought to move to homepage file

darint916 commented on frontend/lyric-sync-app/src/pages/SongPlayerPage/SongPlayerPage.scss:
```diff
@@ -15,13 +20,63 @@
+          
+          .queue-item {
+            &.active-song {
+              background-color: rgba(25, 118, 210, 0.08);
```
variable color ? so we can use color themes in the background? Same with all the other rgba values in the file?

darint916 commented on frontend/lyric-sync-app/src/pages/SongPlayerPage/SongPlayerPage.tsx:
```diff
@@ -25,42 +31,108 @@ const SongPlayerPage: React.FC = () => {
-        // Check if the song is 90% complete
-        if (current / playerRef.current.getDuration() >= 0.9 && isLanguageAllowedForQuiz) {
-          setShowQuizButton(true); // Show the quiz button when 90% complete
+        // Check if the song is 90% complete for quiz button
```
currently working on a pr thats going to do animation frame checks and timestamped checking, much more efficient and precise than the current checking. Can merge into branch. Will update this pr when done

darint916 commented on frontend/lyric-sync-app/src/pages/SongPlayerPage/SongPlayerPage.tsx:
```diff
@@ -71,6 +143,167 @@ const SongPlayerPage: React.FC = () => {
+    });
+  };
+
+  const handleKeyDown = (event: React.KeyboardEvent<HTMLDivElement>) => {
```
isnt this duplicate functions from what i saw in home page, a few things seem duplicate

parthivshaji replied:
Same name, different content

darint916 commented on frontend/lyric-sync-app/src/pages/SongPlayerPage/SongPlayerPage.tsx:
```diff
@@ -71,6 +143,167 @@ const SongPlayerPage: React.FC = () => {
+      
+      // Fallback to basic playback
+      const vidId = extractYouTubeVideoId(item.youTubeId);
+      setCurrentPlayingId(item.queueId);
```
why do we have duplicate code from homepage

parthivshaji replied:
Exported function from homepage and removed duplicate def in songplayerpage

darint916 reviewed (changes requested):
Check the fluff on the code, please make sure you understand every single line you type in and think of efficiency refactoring if possible. For the player itself, I have changes that can be incorporated in soon. Also update the readme + documentation. Also explain how you implemented the queue system. From what I can see, it is quite messy and I think should be in its own component for isolation and ease of understanding. Some code is redundant make sure you can explain why.

darint916 commented:
Running the code myself, I get dependency issue with uuid.
It was not added in the packagelock json. Had to manually install

darint916 commented:
![image](https://github.com/user-attachments/assets/68e2cec5-eb93-4109-a64f-4077579a511f)
Loading too dark, or the text should definitely be contrasted

darint916 commented:
![image](https://github.com/user-attachments/assets/9aca225f-6089-4a3e-8fff-d714cee7537f)
Should definitely get thumbnail placeholders, or a not yet loaded sign so people know if it is loaded or not

darint916 commented:
![image](https://github.com/user-attachments/assets/9b4234ce-3763-4226-92c9-e3fdde896c67)
First song that plays didnt have lyrics but also errored out.
Also getting error playback occurred. Inconsistent highlighting on the first song.

darint916 commented:
Advice from others, for the current text that's playing, add white borders or any to make the current playing line stand out more. The current blue doesn't contrast well with the background
![image](https://github.com/user-attachments/assets/77f1816d-8f67-4d50-83bd-65bba17b8867)

darint916 commented:
getting issue with songs that work on other branches but not on this branch:
Other branches:
![image](https://github.com/user-attachments/assets/d1d16783-18fc-4a03-b025-f7532fd8562f)

Current branch:
![image](https://github.com/user-attachments/assets/6de92171-09db-482f-9f62-b5d61ea12842)
![image](https://github.com/user-attachments/assets/01d47f5f-efaf-433f-ba63-2b7607a9a9f4)
![image](https://github.com/user-attachments/assets/c53d05c1-0b1c-49f7-9789-fc032629d028)

did you send data differently or anything?

parthivshaji commented:
UI update (turned off backend to showcase errors):
![image](https://github.com/user-attachments/assets/ef177365-7ce7-4f55-b42a-a647d4c34b35)

darint916 reviewed (changes requested):
First off, this look so goooood. Messing around with it and it is pretty high quality. Just a few tiny bugs and nitpicks to address. Excellent work otherwise, excited for this.

1.
![image](https://github.com/user-attachments/assets/b1008607-e8ea-4c6c-8123-1961eae2041b)
Small change here, don't like red on red, maybe grey? something else to indicate hover other than same color

2.
![image](https://github.com/user-attachments/assets/c0e0213e-7469-4ec3-9963-5f8e338ad3f7)
I think you should have a separator border other than an abrupt cutoff, either make the Queue tab itself a darker color or have a line like  lyrics:
![image](https://github.com/user-attachments/assets/e8fe89c9-40a3-40a2-b728-438ab86d35ac)

3.
Was wondering if you could add another indicator that something hasn't loaded yet or has loaded. For example:
![image](https://github.com/user-attachments/assets/11fc087c-94aa-40b3-a3f3-1db708fbfe10)
looking at this, you can tell 9 is loaded but 8 and 10 you cant tell, in this case 8 is already loaded. I think for the unloaded ones, you can make it transparent background similar to the way you have it when it is loading like:
![image](https://github.com/user-attachments/assets/d9071c14-7166-412a-b202-ab5b04d4d103)

4.
Was wondering if this loading thing could be put onto the right side without shifting the thumbnail
![image](https://github.com/user-attachments/assets/86e545b4-47c9-4831-9774-25da85cf8175)
so those buffer loading swirls would just simply overwrite the trashcan icon.

5.
Also a nitpick, current song blue dot isn't aligned:
![image](https://github.com/user-attachments/assets/2d41cc82-b867-4d20-aa0c-77ec74cd6f3f)

6.
Also Im not sure if you tested the add song to queue function, tried adding song to queue down there while in a playlist, you didn't lock the search like how you would on the homescreen. I think it should be sent to be attempted to be loaded when you enter into queue and have it locked similar to homescreen, until it loads. If it fails to load, just don't add it. 

7.
Also I forgot to specify, there should be Play Next in queue or Add to End of queue options for the user to specify where to add the song, 
![image](https://github.com/user-attachments/assets/c78bd49e-8fa4-45be-ba9f-99d23955dd2b)
this should be a simple change in making 2 buttons stacked on top aligned with the 2x2 next to it just with those choices, maybe have icons, which are insert to end of stack or next in stack (I think youtube has icons representing those), that should be clear enough

8.
![image](https://github.com/user-attachments/assets/84b561ce-563e-4657-bcdc-33b228fad5d6)

Is there anyway to improve the dragging effect to  not be flickery, I would look into library or youtube music (not too big of concern, could mark as issue for future.

9.
![image](https://github.com/user-attachments/assets/97db2036-b4c0-4cf0-bf17-9e67ddc7d38b)
Think have a number on the top right showing the current song out of how many in queue would be nice. (since people scroll  around in bigger queues)

10.
there is noticeable massive freezelag in queue scrolling when it attempts to find the current place the lyrics (when moving lrc lines) are in LRC. OIr it might be something else? Look into scrolling around in queue, perhaps its because playlist too big? do you do lazy loading? should have lazy loading for massive playlists and load only like 200 songs. Look into the lag issue please. could debug by disabling different parts of the app to see. Do you perhaps have all 3 lrc being rendered? or being rendered at once? should just render the one that has its tab currently open and unmount the component that we aren't viewing? not sure if that is the most optimal, look into it.
currently using : https://www.youtube.com/watch?v=oi8ArKYLwBY&list=PLOqXRy902fv9cStwK80f9ex_9Z0f18DNX&ab_channel=AliAOFFICIAL

11.
![image](https://github.com/user-attachments/assets/0f88eecd-d2cd-4af8-b882-2384d39b0e81)
Need the actual error message please so it is easy to understand why it happened.

!2.
![image](https://github.com/user-attachments/assets/5c12fccd-35b8-4f59-a4b2-938f4e8b1139)
Make errored song clickable still? like load the youtube vid, no need for lyrics. I think some people might have their own videos with just the lyrics.

13,
![image](https://github.com/user-attachments/assets/69570702-8a1a-4d5a-8b44-d907946df3ed)
I get song Id  not defined when clicking quiz. I think we should change the quiz popup button to be much more simple. Less work for you to fix. Basically have the quiz button there the whole time. Can you have 2 buttons that go to each individual quiz instead? Like Go to handwriting quiz, go to Romanized quiz, and it will send the data of the current song to the quiz. or however you want to address it

14. Not sure if it is just me but dragging to a table range outside of queue is really hard and doesn't scroll the table enough. 
![image](https://github.com/user-attachments/assets/5d6927c8-311d-4c30-a9ac-f0056992eb10)
You know how when you drag beyond a certain range near or below the button of the last element of the table, the queue automatically scrolls? try to get that working, lmk if more details are needed to elaborate

15. idea if you want to issue on this
![image](https://github.com/user-attachments/assets/30101e94-3814-4ddf-929b-72b55f750bf1)

you know the spacing between the grids? you should make it be draggable so people can adjust the size of the video to lyric to queue component with min max constraints that we set. Also as a side note, I think you can make queue much more narrow, cuz with it wide, it looks like you unintentionally didn't center the lyrics
![image](https://github.com/user-attachments/assets/0da02adc-6df9-4889-8ad4-f79a84be638b)

16. another thing (you could open issue for this) is displaying the currently playing playlist information in the banner bar (since our nav banner is pretty much blank, maybe just the title and the channel of the playlist (do we even have that info?)) 

17. I think from the homepage you should also pull in the add playlist to queue, if multiple people want to add playlists, should also be there, where it appends theirs to the queue.

18. Shuffle button and repeat button (open issue for this I assume? where it shuffles them)

parthivshaji commented on frontend/lyric-sync-app/src/components/DraggableQueueItem/PlaceholderThumbnail.tsx:
```diff
@@ -0,0 +1,23 @@
+const PlaceholderThumbnail = () => {
+  return (
+    <Box
+      sx={{
```
Done

parthivshaji commented on frontend/lyric-sync-app/src/pages/HomePage/HomePage.tsx:
```diff
@@ -132,7 +128,7 @@ const HomePage: React.FC = () => {
     if (user) {
       try {
-        await axios.post('http://localhost:5259/api/random', 
+        await axios.post('http://localhost:5259/api/random',
```
yessir

parthivshaji commented:
UI update:
![image](https://github.com/user-attachments/assets/59012a0a-cd0a-4cda-8c3f-93472cd66bd3)
![image](https://github.com/user-attachments/assets/9427ec48-9119-4354-ae42-d014e14fe358)

Fixed lagging LRC display.
Fixed quiz bug with songId not being registered.
All suggested styling fixes have been added, except for the drag and drop flickering/scrolling issue, created issue for that #183 
Created issues #184, #185, #186 for recommended features from review.

darint916 reviewed (approved):
n i c e
